File: Backend/Services/PromptBuilder.py
# Builds prompts fromtemplate files in the prompts folder
from pathlib import Path


class PromptBuilder:

    def __init__(self):

        self.promptDirectory = Path("Prompts")

    def LoadTemplate(self, templateName):

        """
        Loads a prompt template from the Prompts folder.
        """

        templatePath = self.promptDirectory / templateName

        try:

            with open(templatePath, "r", encoding="utf-8") as file:

                return file.read()

        except FileNotFoundError:

            raise Exception(
                f"Prompt template '{templateName}' was not found."
            )

    def ReplaceVariables(self, template, variables):

        """
        Replaces placeholders inside a prompt template.

        Example:

        {{profile}}
        """

        completedPrompt = template

        for key, value in variables.items():

            completedPrompt = completedPrompt.replace(
                f"{{{{{key}}}}}",
                str(value)
            )

        return completedPrompt

    def BuildPrompt(self, templateName, variables):

        """
        Loads a template and fills in all placeholders.
        """

        template = self.LoadTemplate(templateName)

        return self.ReplaceVariables(
            template,
            variables
        )

    def BuildResumePrompt(
        self,
        profile,
        experience,
        education,
        skills,
        job
    ):

        variables = {

            "profile": profile,
            "experience": experience,
            "education": education,
            "skills": skills,
            "job": job

        }

        return self.BuildPrompt(
            "Resume.txt",
            variables
        )

    def BuildCoverLetterPrompt(
        self,
        profile,
        experience,
        education,
        skills,
        job
    ):

        variables = {

            "profile": profile,
            "experience": experience,
            "education": education,
            "skills": skills,
            "job": job

        }

        return self.BuildPrompt(
            "CoverLetter.txt",
            variables
        )

    def BuildRewritePrompt(
        self,
        documentType,
        rewriteType,
        content,
        job
    ):

        variables = {

            "documentType": documentType,
            "rewriteType": rewriteType,
            "content": content,
            "job": job

        }

        return self.BuildPrompt(
            "Rewrite.txt",
            variables
        )

File: Backend/Services/promptBuilder.js
// Loads prompt templates from Backend/Prompts and fills in {{placeholders}}.
// These are the same templates the original Python service used.

const fs = require('fs');
const path = require('path');

const PROMPT_DIR = path.join(__dirname, '..', 'Prompts');

function loadTemplate(templateName) {
  const templatePath = path.join(PROMPT_DIR, templateName);

  try {
    return fs.readFileSync(templatePath, 'utf-8');
  } catch (err) {
    throw new Error(`Prompt template '${templateName}' was not found.`);
  }
}

function replaceVariables(template, variables) {
  let result = template;

  for (const [key, value] of Object.entries(variables)) {
    result = result.split(`{{${key}}}`).join(value ?? '');
  }

  return result;
}

function buildPrompt(templateName, variables) {
  const template = loadTemplate(templateName);
  return replaceVariables(template, variables);
}

// Profile fields like skills/education/experience come out of Postgres as
// JSON (arrays of strings or objects). Turn them into readable plain text
// for the prompt instead of dumping raw JSON.
function formatForPrompt(value) {
  if (value === null || value === undefined || value === '') {
    return 'Not provided.';
  }

  if (Array.isArray(value)) {
    if (value.length === 0) return 'Not provided.';

    return value
      .map((item) => {
        if (typeof item === 'string') return `- ${item}`;
        if (item && typeof item === 'object') {
          return `- ${Object.values(item).filter(Boolean).join(' — ')}`;
        }
        return `- ${item}`;
      })
      .join('\n');
  }

  if (typeof value === 'object') {
    return Object.entries(value)
      .map(([k, v]) => `${k}: ${v}`)
      .join('\n');
  }

  return String(value);
}

function buildResumePrompt({ profile, experience, education, skills, job }) {
  return buildPrompt('Resume.txt', {
    profile: formatForPrompt(profile),
    experience: formatForPrompt(experience),
    education: formatForPrompt(education),
    skills: formatForPrompt(skills),
    job: formatForPrompt(job),
  });
}

function buildCoverLetterPrompt({
  profile,
  experience,
  education,
  skills,
  job,
}) {
  return buildPrompt('CoverLetter.txt', {
    profile: formatForPrompt(profile),
    experience: formatForPrompt(experience),
    education: formatForPrompt(education),
    skills: formatForPrompt(skills),
    job: formatForPrompt(job),
  });
}

function buildRewritePrompt({ documentType, rewriteType, content, job }) {
  return buildPrompt('Rewrite.txt', {
    documentType: documentType ?? 'Resume',
    rewriteType: rewriteType ?? 'Professional',
    content: content ?? '',
    job: job ?? 'Not provided.',
  });
}

module.exports = {
  buildResumePrompt,
  buildCoverLetterPrompt,
  buildRewritePrompt,
  formatForPrompt,
};
